nix/nix.go

```diff
@@ -10,15 +10,59 @@ import (
 	"os"
 	"os/exec"
 	"strings"
+
+	"go.jetpack.io/devbox/shell"
 )
 
 func Shell(path string) error {
-	cmd := exec.Command("nix-shell", path)
-	// Default to the shell already being used.
-	shell := os.Getenv("SHELL")
-	if shell != "" {
-		cmd.Args = append(cmd.Args, "--command", shell)
+	// nix-shell only runs bash, which isn't great if the user has a
+	// different default shell. Here we try to detect what their current
+	// shell is, and then `exec` it to replace the bash process inside
+	// nix-shell.
+	sh, err := shell.Detect()
+	if err != nil {
+		// Fall back to running the vanilla Nix bash shell.
+		return runFallbackShell(path)
 	}
+
+	// Naively running the user's shell has two problems:
+	//
+	// 1. The shell will source the user's rc file and potentially reorder
+	// the PATH. This is especially a problem with some shims that prepend
+	// their own directories to the front of the PATH, replacing the
+	// Nix-installed packages.
+	// 2. If their shell is bash, we end up double-sourcing their ~/.bashrc.
+	// Once when nix-shell launches bash, and again when we exec it.
+	//
+	// To workaround this, first we store the current (outside of devbox)
+	// PATH in ORIGINAL_PATH. Then we run a "pure" nix-shell to prevent it
+	// from sourcing their ~/.bashrc. From inside the nix-shell (but before
+	// launching the user's preferred shell) we store the PATH again in
+	// PURE_NIX_PATH. When we're finally in the user's preferred shell, we
+	// can use these env vars to set the PATH so that Nix packages are up
+	// front, and all of the other programs come after.
+	//
+	// ORIGINAL_PATH is set by sh.StartCommand.
+	// PURE_NIX_PATH is set by the shell hook in shell.nix.tmpl.
+	_ = sh.SetInit(`
+# Update the $PATH so the user can keep using programs that live outside of Nix,
+# but prefer anything installed by Nix.
+export PATH="$PURE_NIX_PATH:$ORIGINAL_PATH"
+
+# Prepend to the prompt to make it clear we're in a devbox shell.
+export PS1="(devbox) $PS1"
```

Comment:
Could you add the box emoji too?
```suggestion
export PS1="(📦 devbox) $PS1"
```

Question: Does PS1 work with all the shells?